File: src/main/java/com/snowflake/snowpark_java/UserDefinedFunction.java
package com.snowflake.snowpark_java;

import com.snowflake.snowpark.internal.JavaUtils;

/**
 * Encapsulates a user defined lambda or function that is returned by {@code
 * UDFRegistration.registerTemporary}, {@code UDFRegistration.registerPermanent}, or {@code
 * Functions.udf}.
 *
 * <p>Use {@code UserDefinedFunction.apply} to generate {@code Column} expressions from an instance.
 *
 * <pre>{@code
 * import com.snowflake.snowpark_java.functions;
 * import com.snowflake.snowpark_java.types.DataTypes;
 *
 * UserDefinedFunction myUdf = Functions.udf(
 *   (Integer x, Integer y) -> x + y,
 *   new DataType[]{DataTypes.IntegerType, DataTypes.IntegerType},
 *   DataTypes.IntegerType
 * );
 * df.select(myUdf.apply(df.col("col2"), df.col("col2")));
 * }</pre>
 *
 * @since 0.12.0
 */
public class UserDefinedFunction {
  private final com.snowflake.snowpark.UserDefinedFunction udf;

  UserDefinedFunction(com.snowflake.snowpark.UserDefinedFunction udf) {
    this.udf = udf;
  }

  /**
   * Apply the UDF to one or more columns to generate a Column expression.
   *
   * @since 0.12.0
   * @param exprs The list of expression.
   * @return The result column
   */
  public Column apply(Column... exprs) {
    return new Column(udf.apply(JavaUtils.columnArrayToSeq(Column.toScalaColumnArray(exprs))));
  }
}

File: src/main/scala/com/snowflake/snowpark/UserDefinedFunction.scala
package com.snowflake.snowpark

import com.snowflake.snowpark.internal.ErrorMessage
import com.snowflake.snowpark.internal.analyzer.Expression
import com.snowflake.snowpark.internal.{SnowflakeUDF, UdfColumnSchema}

/** Encapsulates a user defined lambda or function that is returned by
  * [[UDFRegistration.registerTemporary[RT](name* UDFRegistration.registerTemporary]] or by
  * [[com.snowflake.snowpark.functions.udf[RT](* com.snowflake.snowpark.functions.udf]].
  *
  * Use [[UserDefinedFunction!.apply UserDefinedFunction.apply]] to generate [[Column]] expressions
  * from an instance.
  * {{{
  *   import com.snowflake.snowpark.functions._
  *   val myUdf = udf((x: Int, y: String) => y + x)
  *   df.select(myUdf(col("i"), col("s")))
  * }}}
  * @since 0.1.0
  */
case class UserDefinedFunction private[snowpark] (
    f: AnyRef,
    private[snowpark] val returnType: UdfColumnSchema,
    private[snowpark] val inputTypes: Seq[UdfColumnSchema] = Nil,
    name: Option[String] = None
) {

  /** Apply the UDF to one or more columns to generate a [[Column]] expression.
    * @since 0.1.0
    */
  def apply(exprs: Column*): Column = {
    new Column(createUDFExpression(exprs.map(_.expr)))
  }

  private[snowpark] def withName(name: String): UserDefinedFunction = {
    copy(name = Some(name))
  }

  private def createUDFExpression(exprs: Seq[Expression]): SnowflakeUDF = {
    val udfName =
      name.getOrElse(throw ErrorMessage.UDF_FOUND_UNREGISTERED_UDF())
    if (exprs.length != inputTypes.length) {
      throw ErrorMessage.UDF_INCORRECT_ARGS_NUMBER(inputTypes.length, exprs.length)
    }
    SnowflakeUDF(udfName, exprs, returnType.dataType, returnType.isOption)
  }
}
